lib/memo_wise.rb

```diff
@@ -3,6 +3,46 @@
 require "memo_wise/version"
 
 module MemoWise
-  class Error < StandardError; end
-  # Your code goes here...
+  def initialize(*values)
+    @_memo_wise = {}
+    super
+  end
+
+  def self.prepended(target)
+    class << target
+      # Implements memoization for the given method name.
+      #
+      # @param method_name [Symbol]
+      #   Name of method for which to implement memoization.
+      def memo_wise(method_name)
+        not_memoized_name = "#{method_name}_not_memoized".to_sym
+        alias_method not_memoized_name, method_name
+        private not_memoized_name
+
+        module_eval <<-END_OF_METHOD, __FILE__, __LINE__ + 1
+          def #{method_name}(*args)
+            key = ["#{method_name}", args].freeze
+            @_memo_wise.fetch(key) do
+              @_memo_wise[key] = send(:"#{not_memoized_name}", *args)
```

Comment:
We can wait till benchmarking later if you want, but I know already that it will be measurably faster to interpolate the actual method call into this string, instead of meta-programming via `send` -- do we want to change this now?